src/plfa/backmatter/Tips.md

```diff
@@ -0,0 +1,197 @@
+
+## Quickly setting up `with` clauses
+
+If you have a function like this:
+```agda
+foo a = ?
+```
+and you want to introduce a `with`-clause, there is no direct hotkey for this.
+The fastest thing to do is to write
+```agda
+foo a with my-thing
+... | x = ?
+```
+and then ask to case-split on `x`.
+
+## Quickly setting up record skeletons
+
+"Refine" (`C-c C-r` in Emacs) will set up a skeleton for a record constructor if Agda knows a hole must be a record type.
+This is useful for e.g. isomorphisms, products, sigmas, or existentials.
+
+# General tips
+
+## Try and make your definitions compute as much as possible
+
+For example, compare:
+```agda
+2 * foo b
+```
+vs
+```agda
+foo b * 2
+```
+
+The definition of `*` does case analysis on its first argument.
+This means that if the first argument to `*` is a constructor application, Agda can do some computation for you.
+The first example computes to `foo b + (foo b + zero)`, but the second one does not compute at all: Agda can't pattern match on `foo b`, it could be anything!
+
+Doing this lets Agda do more definitional simplification, which means you have to do less work.
+Unfortunately it does require you to know _which_ arguments are the one that Agda can evaluate, which requires looking at the definitions of the functions.
+
+## `x + x` is better than `2 * x`
+
+An expression like `2 * x` will compute to `x + (x + zero)`.
+You now need to get rid of the `x + zero` with a rewrite or similar.
+You don't have this problem if you just write `x + x`.
+
+This is probably less true for larger constants: writing out `8 * x` would be tedious.
+
+## The three styles of proof all tend to use the same assumptions
+
+Consider these three proofs of the same theorem:
+```agda
+*-distrib-+ : ∀ (m n p : ℕ) → (m + n) * p ≡ m * p + n * p
+*-distrib-+ zero n p = refl
+*-distrib-+ (suc m) n p =
+  begin
+    (suc m + n) * p
+  ≡⟨⟩
+    suc (m + n) * p
+  ≡⟨⟩
+    p + (m + n) * p
+  ≡⟨ cong (p +_) (*-distrib-+ m n p) ⟩
+    p + (m * p + n * p)
+  ≡⟨ sym (+-assoc p (m * p) (n * p)) ⟩
+    (p + m * p) + n * p
+  ≡⟨⟩
+    (suc m * p) + n * p
+  ∎
+
+*-distrib-+‵ : ∀ (m n p : ℕ) → (m + n) * p ≡ m * p + n * p
+*-distrib-+‵ zero n p = refl
+*-distrib-+‵ (suc m) n p = trans ( cong (p +_) (*-distrib-+‵ m n p)) ( sym ( +-assoc p (m * p) (n * p)))
+
+*-distrib-+‶ : ∀ (m n p : ℕ) → (m + n) * p ≡ m * p + n * p
+*-distrib-+‶ zero n p = refl
+*-distrib-+‶ (suc m) n p rewrite +-assoc p (m * p) (n * p) | *-distrib-+‶ m n p = refl
+```
+
+They use three approaches for doing "multiple steps":
+1. Equational reasoning
+2. `trans`
+3. `rewrite`
+
+However, notably they all use the same supporting proofs!
+That means you can often write the overall proof however you find easiest and then rewrite it into another form if you want.
+For example, do the proof slowly with equational reasoning, and then turn it into a compact proof with `rewrite`.
+
+## Avoid mutual recursion in proofs by recursing outside the lemma
```

Comment:
I think I understand what this advice is trying to say, but it's a bit unclear. The title just adds to the confusion.

I think what you are trying to say here is that we should try to break up a proof into small lemmas, that do one reasoning step, which we then compose to get the overall goal?